Code/GraphMol/SubstanceGroup.cpp

```diff
@@ -197,6 +296,123 @@ unsigned int addSubstanceGroup(ROMol &mol, SubstanceGroup sgroup) {
   return id;
 }
 
+namespace {
+bool includesBond(SubstanceGroup &sg, unsigned int idx) {
+  return sg.includesBond(idx);
+}
+bool includesAtom(SubstanceGroup &sg, unsigned int idx) {
+  return sg.includesAtom(idx);
+}
+void removedBond(SubstanceGroup &sg, unsigned int idx) {
+  sg.adjustToRemovedBond(idx);
+}
+void removedAtom(SubstanceGroup &sg, unsigned int idx) {
+  sg.adjustToRemovedAtom(idx);
+}
+
+bool removedParentInHierarchy(
+    unsigned int idx, const std::vector<SubstanceGroup> &sgs,
+    const boost::dynamic_bitset<> &toRemove,
+    const std::map<unsigned int, unsigned int> &indexLookup) {
+  PRECONDITION(idx < sgs.size(), "cannot find SubstanceGroup");
+  if (toRemove[idx]) return true;
+
+  unsigned int parent;
+  if (sgs[idx].getPropIfPresent("PARENT", parent)) {
+    auto piter = indexLookup.find(parent);
+    if (piter != indexLookup.end()) {
+      return removedParentInHierarchy(piter->second, sgs, toRemove,
+                                      indexLookup);
+    }
+  }
+  return false;
+}
+
+template <bool INCLUDES_METHOD(SubstanceGroup &, unsigned int),
+          void ADJUST_METHOD(SubstanceGroup &, unsigned int)>
+void removeSubstanceGroupsReferencing(RWMol &mol, unsigned int idx) {
+  auto &sgs = getSubstanceGroups(mol);
+  if (!sgs.empty()) {
+    // first collect the ones that should be removed
+    boost::dynamic_bitset<> toRemove(sgs.size());
+    unsigned int nRemoved = 0;
+    bool parentsPresent = false;
+    for (unsigned int i = 0; i < sgs.size(); ++i) {
+      if (!parentsPresent && sgs[i].hasProp("PARENT")) {
+        parentsPresent = true;
+      }
+      if (INCLUDES_METHOD(sgs[i], idx)) {
+        toRemove.set(i);
+        ++nRemoved;
+      }
+    }
+
+    // if we're going to be removing anything and there are PARENTS present,
+    // we need to build a lookup map between index->position in original array
+    std::map<unsigned int, unsigned int> indexLookup;
```

Comment:
Maybe use an `unordered_map`?

Reply:
What's the benefit?

Reply:
Not really too much, as indexes will ints, and we won't have many of them. I guess I am biased against map.